cilium/CFP-30984-dns-proxy-ha.md

```diff
@@ -0,0 +1,231 @@
+
+##### Pros
+
+* Simpler to maintain and stay upto date with any new ip<>identity changes.
+* All data can be cached in-memory, in an event of agent being unavailable, SDP can lookup from in-memory cache
+
+##### Cons
+
+* Need a mechanism for reconciling the cache in case of SDP restarts.
+* SDP will be aware of all the endpoints data, which might not be needed for DNS proxy.
+
+#### Option 2: Read from bpf maps
+
+Read mappings from bpf maps `cilium_lxc` for endpoint ID and `cilium_ipcache` for identity.
+
+##### Pros
+
+* Reduced resource utilization in agent since agent doesn't need to process rpc calls in hotpath
+* IP to endpoint ID or identity translation does not rely on agent availability.
+* DNS can be resolved even when agent is down.
+
+##### Cons
+
+* Low level interactions from components external to cilium agent.
+
+### Flag to disable built-in proxy
+
+Adding a flag in cilium agent to disable built-in dns proxy allows for de-coupling of SDP's lifecycle from agent. The current implementation of cilium agent depends on the built-in dns proxy for restoring the endpoint for DNS rules. With the flag enabled we will need to refactor to remove that dependency. Cilium agent can retrieve the DNS rules from in memory policy representation (agent already parses this information from ep_config.json on startup).
+
+#### Pros
+
+* Allows for de-coupling upgrade cycles of SDP and cilium agent.
+* Allows any delegated DNS proxy to be plugged in.
+* Reduces the resource utilization in cilium agent.
+
+#### Cons
+
+* Higher latency in DNS resolution as the DNS queries are always handled by a separate process.
+
+### Q: Restoring L7 DNS policy when agent is unavailable
+
+In an event where SDP restarts when agent's gRPC service is unavailable, SDP can read state from `ep_config.json` and restore policy.
+
+### Indentity for upstream dns pods getting updated
+
+We are sending the set of identities for upstream dns pods along with the dns rules to the SDP through grpc stream. In case the upstream dns pods labels are updated, it will trigger the policy recalculation and the new set of identities will be sent to the SDP. This will help in enforcing the L7 DNS policy for the updated set of identities.
+
+### Scenarios and Expected Behaviour
+
+In an ideal scenario, both cilium agent and SDP should be interacting with each other through a biredirectional grpc stream. However, in an event where either of them restarts/upgrade/downgrade, what should be the expected behavior ?
+
+#### Cilium Agent is down, SDP is running
+
+* SDP should be able to proxy requests and enfore L7 DNS policy based on the existing bpf maps and in memory DNS rules. The datapath of already configured policies will work. Any new mappings will not be updated until cilium agent is back up.
```

Comment:
Just to clarify the mappings refers to the new policy map updates.
I think it follows the same pattern as it is now in case of cilium agent not available. The policy won't be applied until cilium agent is up and running.

Reply:
OK, how does the design ensure eventual consistency on that front once the agent comes back up?

Reply:
SDP retries the connection to cilium agent if there is a connection failure to cilium agent. I think a controller (a separate go routine) should be able keep a track of an active connection with cilium agent and retrigger connection request in it is broken.

Reply:
Sounds good, I think it makes sense to integrate this into the text and just highlight that anything more is up to implementation details.